### PR TITLE
NodeOperatorRegistry limits features v2

### PR DESCRIPTION
**Pull request by loga4**
- added stuck/stuck time/forgiven counters
- added target limit
- added `invalidateReadyToDepositKeysRange(uint256 _indexFrom, uint256 _indexTo)`
- reduce roles
- rewrite key management

codesize: `24.521`

### CODE REVIEW

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol [hunk 1]
```diff
@@ -507,77 +699,86 @@ contract NodeOperatorsRegistry is AragonApp, IStakingModule, Versioned {
...
+
+        Packed64x4.Packed memory operatorTargetStats = _loadOperatorTargetValidtatorsStats(_nodeOperatorId);
+        // correct vetted count according to target
+        if (operatorTargetStats.get(TARGET_VALIDATORS_ACTIVE_OFFSET) > 0) {
+            // if target is enabled
+            uint64 targetValidatorsCount = operatorTargetStats.get(TARGET_VALIDATORS_COUNT_OFFSET);
+            if (targetValidatorsCount <= depositedSigningKeysCount.sub(exitedSigningKeysCount)) {
+                vettedSigningKeysCount = depositedSigningKeysCount;
+            } else {
+                vettedSigningKeysCount = Math64.min(vettedSigningKeysCount, exitedSigningKeysCount.add(targetValidatorsCount));
```
Comment (loga4) [inline contracts/0.4.24/nos/NodeOperatorsRegistry.sol:746]: why `exitedSigningKeysCount+targetValidatorsCount`? and not just `targetValidatorsCount`?

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol [hunk 2]
```diff
@@ -507,77 +699,86 @@ contract NodeOperatorsRegistry is AragonApp, IStakingModule, Versioned {
...
+        if (!getNodeOperatorIsActive(_nodeOperatorId)) return (0, 0, 0);
+
+        Packed64x4.Packed memory signingKeysStats = _loadOperatorSigningKeysStats(_nodeOperatorId);
+        vettedSigningKeysCount = signingKeysStats.get(VETTED_KEYS_COUNT_OFFSET);
+        exitedSigningKeysCount = signingKeysStats.get(EXITED_KEYS_COUNT_OFFSET);
+        depositedSigningKeysCount = signingKeysStats.get(DEPOSITED_KEYS_COUNT_OFFSET);
+
+        Packed64x4.Packed memory operatorTargetStats = _loadOperatorTargetValidtatorsStats(_nodeOperatorId);
+        // correct vetted count according to target
+        if (operatorTargetStats.get(TARGET_VALIDATORS_ACTIVE_OFFSET) > 0) {
```
Comment (loga4) [inline contracts/0.4.24/nos/NodeOperatorsRegistry.sol:740]: we can set target to 0, maybe check by >= ?

File: contracts/0.4.24/lib/Packed64x4.sol [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    //extract n-th uint64 from uint256
+    function get(Packed memory _self, uint8 n) internal pure returns (uint64 r) {
+        // require(n < 4, ERROR_OFFSET_OUT_OF_RANGE);
+        r = uint64((_self.v >> (64 * n)) & 0xFFFFFFFFFFFFFFFF);
+    }
+
+    //merge n-th uint64 to uint256
+    function set(Packed memory _self, uint8 n, uint64 x) internal pure {
+        // require(n < 4, ERROR_OFFSET_OUT_OF_RANGE);
+        _self.v = _self.v & ~(uint256(0xFFFFFFFFFFFFFFFF) << (64 * n)) | ((uint256(x) & 0xFFFFFFFFFFFFFFFF) << (64 * n));
```
Comment (TheDZhon) [inline contracts/0.4.24/lib/Packed64x4.sol:31]: can it be?
```solidity
_self.v &= ~(uint256(0xFFFFFFFFFFFFFFFF) << (64 * n)) | (uint256(x) << (64 * n));
```

File: contracts/0.4.24/lib/SigningKeys.sol [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    event SigningKeyAdded(uint256 indexed nodeOperatorId, bytes pubkey);
+    event SigningKeyRemoved(uint256 indexed nodeOperatorId, bytes pubkey);
+
+    function isKeyEmpty(bytes memory _key) internal pure returns (bool) {
+        assert(_key.length == PUBKEY_LENGTH);
+
+        uint256 k1;
+        uint256 k2;
+        assembly {
+            k1 := mload(add(_key, 0x20))
```
Comment (TheDZhon) [inline contracts/0.4.24/lib/SigningKeys.sol:32]: DISCLAIMER :skull: 
totally not sure that this is a good idea, but decided to try my best :crab: 
```
        bool ret;
        assembly {
            ret  := iszero(and(mload(add(_key, 0x20)), shr(mload(add(_key, 0x40)), 0x80)))
        }
 ```

File: contracts/0.4.24/lib/SigningKeys.sol [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+            k2 := mload(add(_key, 0x40))
+        }
+
+        return 0 == k1 && 0 == (k2 >> ((2 * 32 - PUBKEY_LENGTH) * 8));
```
Comment (TheDZhon) [inline contracts/0.4.24/lib/SigningKeys.sol:36]: DISCLAIMER :skull: 
totally not sure that this is a good idea, but decided to try my best :crab: 
```
        bool ret;
        assembly {
            ret  := iszero(and(mload(add(_key, 0x20)), shr(mload(add(_key, 0x40)), 0x80)))
        }
 ```

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol [hunk 6]
```diff
@@ -12,22 +12,28 @@ import {UnstructuredStorage} from "@aragon/os/contracts/common/UnstructuredStora
...
-import {SigningKeysStats} from "../lib/SigningKeysStats.sol";
+import {ILidoLocator} from "../../common/interfaces/ILidoLocator.sol";
+import {SigningKeys} from "../lib/SigningKeys.sol";
+import {Packed64x4} from "../lib/Packed64x4.sol";
 import {Versioned} from "../utils/Versioned.sol";
 
 interface IStETH {
     function sharesOf(address _account) external view returns (uint256);
     function transferShares(address _recipient, uint256 _sharesAmount) external returns (uint256);
+    function burnShares(address _account, uint256 _sharesAmount) external returns (uint256 newTotalShares);
```
Comment (TheDZhon) [inline contracts/0.4.24/nos/NodeOperatorsRegistry.sol:23]: it has been removed already

please use `IBurner.requestBurnShares(adddress _account, uint256 _sharesAmount)` as a drop-in replacement

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol [hunk 7]
```diff
@@ -12,22 +12,28 @@ import {UnstructuredStorage} from "@aragon/os/contracts/common/UnstructuredStora
...
 
 /// @dev This interface describes only tiny part of the full interface, which NodeOperatorsRegistry must implement
 ///      See 0.8.9/interface/IStakingModule.sol for the full version.
 ///      We don't inherit 0.8.9 IStakingModule due to the solidity version conflict.
 interface IStakingModule {
+    // function updateStuckValidatorsKeysCount(uint256 nodeOperatorId, uint64 stuckValidatorsCount) external;
+
     event ValidatorsKeysNonceChanged(uint256 validatorsKeysNonce);
+    event StuckValidatorsCountChanged(uint256 indexed nodeOperatorId, uint256 stuckValidatorsCount);
+    event ForgivenValidatorsCountChanged(uint256 indexed nodeOperatorId, uint256 forgivenValidatorsCount);
```
Comment (TheDZhon) [inline contracts/0.4.24/nos/NodeOperatorsRegistry.sol:34]: `Refunded`

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol [hunk 8]
```diff
@@ -1047,136 +1252,108 @@ contract NodeOperatorsRegistry is AragonApp, IStakingModule, Versioned {
...
 
         distributed = 0;
+
         for (uint256 idx = 0; idx < recipients.length; ++idx) {
             if (shares[idx] == 0) continue;
+            if (penalized[idx]) {
+                /// @dev half reward punishment
+                /// @dev ignore remainder since it accumulated on contract balance
+                shares[idx] >>= 1;
+                stETH.transferShares(getLocator().burner(), shares[idx]);
```
Comment (TheDZhon) [inline contracts/0.4.24/nos/NodeOperatorsRegistry.sol:1273]:
```solidity
IBurner(getLocator().burner()).requestBurnShares(address(this), shares[idx]);
```
allowance needed (maybe ∞) from NOR to Burner

**Review by TheDZhon — COMMENTED** [review]
:eyes:

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol (hunk 2 above)
Comment (loga4) [reply contracts/0.4.24/nos/NodeOperatorsRegistry.sol:740]: it's correct

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol (hunk 1 above)
Comment (loga4) [reply contracts/0.4.24/nos/NodeOperatorsRegistry.sol:746]: it's correct. 
targetValidatorsCount - is about active keys

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol (hunk 7 above)
Comment (loga4) [reply contracts/0.4.24/nos/NodeOperatorsRegistry.sol]: fixed

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol [hunk 9]
```diff
@@ -12,22 +12,29 @@ import {UnstructuredStorage} from "@aragon/os/contracts/common/UnstructuredStora
...
 interface IStETH {
     function sharesOf(address _account) external view returns (uint256);
     function transferShares(address _recipient, uint256 _sharesAmount) external returns (uint256);
+    function approve(address _spender, uint256 _amount) external returns (bool);
 }
 
 /// @dev This interface describes only tiny part of the full interface, which NodeOperatorsRegistry must implement
 ///      See 0.8.9/interface/IStakingModule.sol for the full version.
 ///      We don't inherit 0.8.9 IStakingModule due to the solidity version conflict.
 interface IStakingModule {
```
Comment (ujenjt) [inline contracts/0.4.24/nos/NodeOperatorsRegistry.sol:30]: let's remove

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol (hunk 8 above)
Comment (krogla) [reply contracts/0.4.24/nos/NodeOperatorsRegistry.sol]: done

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol (hunk 6 above)
Comment (krogla) [reply contracts/0.4.24/nos/NodeOperatorsRegistry.sol]: yep

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol (hunk 9 above)
Comment (krogla) [reply contracts/0.4.24/nos/NodeOperatorsRegistry.sol]: done

File: contracts/0.4.24/lib/Packed64x4.sol (hunk 3 above)
Comment (krogla) [reply contracts/0.4.24/lib/Packed64x4.sol:31]: yes, but actually no, i already tried 
idk

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol [hunk 10]
```diff
@@ -49,62 +47,103 @@ contract NodeOperatorsRegistry is AragonApp, IStakingModule, Versioned {
...
     // CONSTANTS
     //
     uint256 public constant MAX_NODE_OPERATORS_COUNT = 200;
     uint256 public constant MAX_NODE_OPERATOR_NAME_LENGTH = 255;
 
-    uint64 private constant PUBKEY_LENGTH = 48;
-    uint64 private constant SIGNATURE_LENGTH = 96;
-    uint256 internal constant UINT64_MAX = uint256(uint64(-1));
+    uint256 internal constant UINT64_MAX = uint256(~uint64(0));
+    uint256 internal constant STUCK_VALIDATORS_PENALTY_DELAY = 2 days;
```
Comment (ujenjt) [inline contracts/0.4.24/nos/NodeOperatorsRegistry.sol:90]: let's convert to getter/setter

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol [hunk 11]
```diff
@@ -49,62 +47,103 @@ contract NodeOperatorsRegistry is AragonApp, IStakingModule, Versioned {
...
+        0xeaf200990dc6840b2b98dda560b1fd49fc2bbb13aae6f2864e84b7af0b3026fe;
+    // bytes32 public constant ACTIVATE_NODE_OPERATOR_ROLE = keccak256("MANAGE_NODE_OPERATOR_ROLE");
+    bytes32 public constant MANAGE_NODE_OPERATOR_ROLE = 0x78523850fdd761612f46e844cf5a16bda6b3151d6ae961fd7e8e7b92bfbca7f8;
+    // bytes32 public constant STAKING_ROUTER_ROLE = keccak256("STAKING_ROUTER_ROLE");
+    bytes32 public constant STAKING_ROUTER_ROLE = 0xbb75b874360e0bfd87f964eadd8276d8efb7c942134fc329b513032d0803e0c6;
 
     //
     // CONSTANTS
     //
     uint256 public constant MAX_NODE_OPERATORS_COUNT = 200;
```
Comment (ujenjt) [inline contracts/0.4.24/nos/NodeOperatorsRegistry.sol:86]: What is safe limit now? still 200?

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol [hunk 12]
```diff
@@ -49,62 +47,103 @@ contract NodeOperatorsRegistry is AragonApp, IStakingModule, Versioned {
...
     uint256 public constant MAX_NODE_OPERATOR_NAME_LENGTH = 255;
 
-    uint64 private constant PUBKEY_LENGTH = 48;
-    uint64 private constant SIGNATURE_LENGTH = 96;
-    uint256 internal constant UINT64_MAX = uint256(uint64(-1));
+    uint256 internal constant UINT64_MAX = uint256(~uint64(0));
```
Comment (ujenjt) [inline contracts/0.4.24/nos/NodeOperatorsRegistry.sol:89]: 0xff...

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol [hunk 13]
```diff
@@ -296,9 +319,8 @@ contract NodeOperatorsRegistry is AragonApp, IStakingModule, Versioned {
     function setNodeOperatorName(uint256 _nodeOperatorId, string _name) external {
         _onlyValidNodeOperatorName(_name);
         _onlyExistedNodeOperator(_nodeOperatorId);
-        _authP(SET_NODE_OPERATOR_NAME_ROLE, arr(uint256(_nodeOperatorId)));
+        _authP(ADD_NODE_OPERATOR_ROLE, arr(uint256(_nodeOperatorId)));
 
-        require(keccak256(_nodeOperators[_nodeOperatorId].name) != keccak256(_name), "NODE_OPERATOR_NAME_IS_THE_SAME");
```
Comment (ujenjt) [inline contracts/0.4.24/nos/NodeOperatorsRegistry.sol:301]: let's revert this change since we have some space for that

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol [hunk 14]
```diff
@@ -49,62 +47,103 @@ contract NodeOperatorsRegistry is AragonApp, IStakingModule, Versioned {
...
+    bytes32 public constant MANAGE_SIGNING_KEYS = 0x75abc64490e17b40ea1e66691c3eb493647b24430b358bd87ec3e5127f1621ee;
+    // bytes32 public constant ADD_NODE_OPERATOR_ROLE = keccak256("ADD_NODE_OPERATOR_ROLE");
+    bytes32 public constant ADD_NODE_OPERATOR_ROLE = 0xe9367af2d321a2fc8d9c8f1e67f0fc1e2adf2f9844fb89ffa212619c713685b2;
+    // bytes32 public constant SET_NODE_OPERATOR_LIMIT_ROLE = keccak256("SET_NODE_OPERATOR_LIMIT_ROLE");
+    bytes32 public constant SET_NODE_OPERATOR_LIMIT_ROLE = 0x07b39e0faf2521001ae4e58cb9ffd3840a63e205d288dc9c93c3774f0d794754;
+    // bytes32 public constant INVALIDATE_READY_TO_DEPOSIT_KEYS_ROLE = keccak256("INVALIDATE_READY_TO_DEPOSIT_KEYS_ROLE");
+    bytes32 public constant INVALIDATE_READY_TO_DEPOSIT_KEYS_ROLE =
+        0xeaf200990dc6840b2b98dda560b1fd49fc2bbb13aae6f2864e84b7af0b3026fe;
+    // bytes32 public constant ACTIVATE_NODE_OPERATOR_ROLE = keccak256("MANAGE_NODE_OPERATOR_ROLE");
+    bytes32 public constant MANAGE_NODE_OPERATOR_ROLE = 0x78523850fdd761612f46e844cf5a16bda6b3151d6ae961fd7e8e7b92bfbca7f8;
```
Comment (ujenjt) [inline contracts/0.4.24/nos/NodeOperatorsRegistry.sol:79]: let's combine w/ ADD_NODE_OPERATOR_ROLE

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol [hunk 15]
```diff
@@ -412,90 +422,128 @@ contract NodeOperatorsRegistry is AragonApp, IStakingModule, Versioned {
...
+        returns (uint64)
     {
-        _updateStuckValidatorsKeysCount(_nodeOperatorId, _stuckValidatorsKeysCount, true);
-        _updateExitedValidatorsKeysCount(_nodeOperatorId, _exitedValidatorsKeysCount, true);
+        Packed64x4.Packed memory signingKeysStats = _loadOperatorSigningKeysStats(_nodeOperatorId);
+        uint64 depositedSigningKeysCount = signingKeysStats.get(DEPOSITED_KEYS_COUNT_OFFSET);
+        uint64 exitedValidatorsCount = signingKeysStats.get(EXITED_KEYS_COUNT_OFFSET);
+    
+        if (exitedValidatorsCount != _exitedValidatorsKeysCount) {
+            _requireOutOfRange(_exitedValidatorsKeysCount <= depositedSigningKeysCount);
```
Comment (ujenjt) [inline contracts/0.4.24/nos/NodeOperatorsRegistry.sol:442]: _requireValidRange

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol [hunk 16]
```diff
@@ -412,90 +422,128 @@ contract NodeOperatorsRegistry is AragonApp, IStakingModule, Versioned {
...
+            if (_exitedValidatorsKeysCount < exitedValidatorsCount) {
```
Comment (ujenjt) [inline contracts/0.4.24/nos/NodeOperatorsRegistry.sol:443]:
```
if (!_allowDecrease &&_exitedValidatorsKeysCount >= exitedValidatorsCount) {
   throw
}
```

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol [hunk 17]
```diff
@@ -412,90 +422,128 @@ contract NodeOperatorsRegistry is AragonApp, IStakingModule, Versioned {
...
+                require(_allowDecrease, "EXITED_VALIDATORS_COUNT_DECREASED");
+            }
+
+            signingKeysStats.set(EXITED_KEYS_COUNT_OFFSET, _exitedValidatorsKeysCount);
+            _saveOperatorSigningKeysStats(_nodeOperatorId, signingKeysStats);
+            emit ExitedSigningKeysCountChanged(_nodeOperatorId, _exitedValidatorsKeysCount);
+            return _exitedValidatorsKeysCount;
```
Comment (ujenjt) [inline contracts/0.4.24/nos/NodeOperatorsRegistry.sol:450]: fix staking router

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol [hunk 18]
```diff
@@ -412,90 +422,128 @@ contract NodeOperatorsRegistry is AragonApp, IStakingModule, Versioned {
...
-    function _updateExitedValidatorsKeysCount(
-        uint256 _nodeOperatorId,
-        uint256 _exitedValidatorsKeysCount,
-        bool _allowDecrease
-    ) internal returns (uint256) {
+    /// @notice Updates the limit of the validators that can be used for deposit by DAO
+    /// @param _nodeOperatorId Id of the node operator
+    /// @param _targetValidatorsCount New number of EXITED validators of the node operator
+    /// @param _targetActive active flag
+    function updateTargetValidatorsLimits(uint256 _nodeOperatorId, uint64 _targetValidatorsCount, bool _targetActive)
```
Comment (ujenjt) [inline contracts/0.4.24/nos/NodeOperatorsRegistry.sol:460]: function updateTargetValidatorsLimits(uint256 _nodeOperatorId, bool _isTargetLimitActive, uint256 _targetLimit)

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol [hunk 19]
```diff
@@ -387,14 +396,15 @@ contract NodeOperatorsRegistry is AragonApp, IStakingModule, Versioned {
         external
         returns (uint256)
     {
-        return _updateExitedValidatorsKeysCount(_nodeOperatorId, _exitedValidatorsKeysCount, false);
+        _onlyExistedNodeOperator(_nodeOperatorId);
+        _auth(STAKING_ROUTER_ROLE);
+
+        return _updateExitedValidatorsKeysCount(_nodeOperatorId, uint64(_exitedValidatorsKeysCount), false);
```
Comment (ujenjt) [inline contracts/0.4.24/nos/NodeOperatorsRegistry.sol:402]: if (stuckPenaltyStats.get(STUCK_VALIDATORS_COUNT_OFFSET) <= _refundedValidatorsCount) {
            stuckPenaltyStats.set(STUCK_PENALTY_END_TIMESTAMP_OFFSET, uint64(block.timestamp + STUCK_VALIDATORS_PENALTY_DELAY));
        }

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol [hunk 20]
```diff
@@ -370,11 +378,12 @@ contract NodeOperatorsRegistry is AragonApp, IStakingModule, Versioned {
...
-    function updateStuckValidatorsKeysCount(uint256 _nodeOperatorId, uint256 _stuckValidatorKeysCount)
-        external
-    {
-        _updateStuckValidatorsKeysCount(_nodeOperatorId, _stuckValidatorKeysCount, false);
+    /// @param _stuckValidatorsCount New number of stuck validators of the node operator
+    function updateStuckValidatorsKeysCount(uint256 _nodeOperatorId, uint256 _stuckValidatorsCount) external {
+        _onlyExistedNodeOperator(_nodeOperatorId);
+        _auth(STAKING_ROUTER_ROLE);
+
+        _updateStuckValidatorsKeysCount(_nodeOperatorId, uint64(_stuckValidatorsCount));
```
Comment (ujenjt) [inline contracts/0.4.24/nos/NodeOperatorsRegistry.sol:386]: if (stuckPenaltyStats.get(STUCK_VALIDATORS_COUNT_OFFSET) <= _refundedValidatorsCount) {
            stuckPenaltyStats.set(STUCK_PENALTY_END_TIMESTAMP_OFFSET, uint64(block.timestamp + STUCK_VALIDATORS_PENALTY_DELAY));
        }

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol [hunk 21]
```diff
@@ -412,90 +422,128 @@ contract NodeOperatorsRegistry is AragonApp, IStakingModule, Versioned {
...
 
-    function _updateStuckValidatorsKeysCount(
-        uint256 _nodeOperatorId,
-        uint256 _stuckValidatorKeysCount,
-        bool _allowDecrease
-    ) internal {
-        // FIXME: implement
+    function invalidateReadyToDepositKeys() external {
+        uint256 operatorsCount = getNodeOperatorsCount();
+        _requireOutOfRange(operatorsCount > 0);
```
Comment (ujenjt) [inline contracts/0.4.24/nos/NodeOperatorsRegistry.sol:515]: let's remove

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol [hunk 22]
```diff
@@ -507,77 +555,83 @@ contract NodeOperatorsRegistry is AragonApp, IStakingModule, Versioned {
...
+            enqueuedValidatorsKeysCount, nodeOperatorIds, activeKeysCountAfterAllocation, exitedSigningKeysCount
         );
 
-        SigningKeysStats.State memory totalSigningKeysStats = _getTotalSigningKeysStats();
-        totalSigningKeysStats.increaseDepositedSigningKeysCount(uint64(enqueuedValidatorsKeysCount));
-        _setTotalSigningKeysStats(totalSigningKeysStats);
         _increaseValidatorsKeysNonce();
     }
 
+    function _getCorrectedNodeOperator(uint256 _nodeOperatorId)
```
Comment (ujenjt) [inline contracts/0.4.24/nos/NodeOperatorsRegistry.sol:579]: _getNodeOperatorWithLimitApplied(uint256 _nodeOperatorId)

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol [hunk 23]
```diff
@@ -507,77 +555,83 @@ contract NodeOperatorsRegistry is AragonApp, IStakingModule, Versioned {
...
-            NodeOperator storage nodeOperator = _nodeOperators[nodeOperatorId];
-            if (!nodeOperator.active) continue;
-
-            uint256 depositedSigningKeysCount = nodeOperator.depositedSigningKeysCount;
-            uint256 vettedSigningKeysCount = nodeOperator.vettedSigningKeysCount;
-
+            (vettedSigningKeysCount, exitedSigningKeysCount, depositedSigningKeysCount) =
+                _getCorrectedNodeOperator(nodeOperatorId);
             // the node operator has no available signing keys
             if (depositedSigningKeysCount == vettedSigningKeysCount) continue;
```
Comment (ujenjt) [inline contracts/0.4.24/nos/NodeOperatorsRegistry.sol:630]: check that this NO are not penalized

Comment (ujenjt) [reply contracts/0.4.24/nos/NodeOperatorsRegistry.sol:630]: 1) stucked > refunded
2) no timeouts

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol [hunk 24]
```diff
@@ -780,116 +826,95 @@ contract NodeOperatorsRegistry is AragonApp, IStakingModule, Versioned {
...
-    function _removeUnusedSigningKeys(
-        uint256 _nodeOperatorId,
-        uint64 _fromIndex,
-        uint64 _keysCount
-    ) internal {
+    function _removeUnusedSigningKeys(uint256 _nodeOperatorId, uint256 _fromIndex, uint256 _keysCount) internal {
         _onlyExistedNodeOperator(_nodeOperatorId);
         _onlyNodeOperatorManager(msg.sender, _nodeOperatorId);
 
+        _requireOutOfRange(_fromIndex < UINT64_MAX);
```
Comment (ujenjt) [inline contracts/0.4.24/nos/NodeOperatorsRegistry.sol:860]: Think how to optimimize checks, maybe like this:

require(_fromIndex.add(_keysCount) <= totalSigningKeysCount, "KEY_NOT_FOUND");

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol [hunk 25]
```diff
@@ -780,116 +826,95 @@ contract NodeOperatorsRegistry is AragonApp, IStakingModule, Versioned {
...
+        return signingKeysStats.get(TOTAL_KEYS_COUNT_OFFSET);
     }
 
     /// @notice Returns number of usable signing keys of the node operator #`_nodeOperatorId`
     function getUnusedSigningKeyCount(uint256 _nodeOperatorId) external view returns (uint256) {
         _onlyExistedNodeOperator(_nodeOperatorId);
-        return _nodeOperators[_nodeOperatorId].totalSigningKeysCount.sub(_nodeOperators[_nodeOperatorId].depositedSigningKeysCount);
+
+        Packed64x4.Packed memory signingKeysStats = _loadOperatorSigningKeysStats(_nodeOperatorId);
+        return signingKeysStats.diff(TOTAL_KEYS_COUNT_OFFSET, DEPOSITED_KEYS_COUNT_OFFSET);
```
Comment (ujenjt) [inline contracts/0.4.24/nos/NodeOperatorsRegistry.sol:907]: lets' uwrap and than sub

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol [hunk 26]
```diff
@@ -932,92 +940,122 @@ contract NodeOperatorsRegistry is AragonApp, IStakingModule, Versioned {
...
-    /// @return readyToDepositValidatorsKeysCount Total number of validators ready to be deposited
-    function getValidatorsKeysStats()
+    function getNodeOperatorStats(uint256 _nodeOperatorId)
         external
         view
         returns (
-            uint256 exitedValidatorsCount,
-            uint256 activeValidatorsKeysCount,
-            uint256 readyToDepositValidatorsKeysCount
+            bool targetValidatorsActive,
```
Comment (ujenjt) [inline contracts/0.4.24/nos/NodeOperatorsRegistry.sol:972]: targetLimitActive

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol [hunk 27]
```diff
@@ -655,48 +711,64 @@ contract NodeOperatorsRegistry is AragonApp, IStakingModule, Versioned {
...
+            Packed64x4.Packed memory signingKeysStats = _loadOperatorSigningKeysStats(operatorId);
+            uint256 activeValidatorsCount = signingKeysStats.diff(DEPOSITED_KEYS_COUNT_OFFSET, EXITED_KEYS_COUNT_OFFSET);
             totalActiveValidatorsCount = totalActiveValidatorsCount.add(activeValidatorsCount);
 
-            recipients[idx] = nodeOperator.rewardAddress;
+            recipients[idx] = _nodeOperators[operatorId].rewardAddress;
+            // prefill shares array with 'key share' for recipient, see below
             shares[idx] = activeValidatorsCount;
 
+            Packed64x4.Packed memory stuckPenaltyStats = _loadOperatorStuckPenaltyStats(operatorId);
```
Comment (ujenjt) [inline contracts/0.4.24/nos/NodeOperatorsRegistry.sol:752]: _isOperatorPenalized

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol [hunk 28]
```diff
@@ -932,92 +940,122 @@ contract NodeOperatorsRegistry is AragonApp, IStakingModule, Versioned {
...
+        targetValidatorsActive = operatorTargetStats.get(TARGET_VALIDATORS_ACTIVE_OFFSET) != 0;
+        targetValidatorsCount = operatorTargetStats.get(TARGET_VALIDATORS_COUNT_OFFSET);
+        stuckValidatorsCount = stuckPenaltyStats.get(STUCK_VALIDATORS_COUNT_OFFSET);
+        refundedValidatorsCount = stuckPenaltyStats.get(REFUNDED_VALIDATORS_COUNT_OFFSET);
+        stuckPenaltyEndTimestamp = stuckPenaltyStats.get(STUCK_PENALTY_END_TIMESTAMP_OFFSET);
     }
 
+   
+
+    function _isOperatorPenalized(uint256 _nodeOperatorId) internal view returns (bool) {
```
Comment (ujenjt) [inline contracts/0.4.24/nos/NodeOperatorsRegistry.sol:993]: public

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol [hunk 29]
```diff
@@ -932,92 +940,122 @@ contract NodeOperatorsRegistry is AragonApp, IStakingModule, Versioned {
...
+        uint256 vettedSigningKeysCount = signingKeysStats.get(VETTED_KEYS_COUNT_OFFSET);
+
+        exitedValidatorsCount = signingKeysStats.get(EXITED_KEYS_COUNT_OFFSET);
+        activeValidatorsKeysCount = depositedSigningKeysCount - exitedValidatorsCount;
+
+        if (!_isOperatorPenalized(_nodeOperatorId)) {
+            Packed64x4.Packed memory operatorTargetStats = _loadOperatorTargetValidatorsStats(_nodeOperatorId);
+            uint256 targetLimit = exitedValidatorsCount + operatorTargetStats.get(TARGET_VALIDATORS_COUNT_OFFSET);
+            // if target moede is active and staking limit is greater than targeted keys amount
+            if (operatorTargetStats.get(TARGET_VALIDATORS_ACTIVE_OFFSET) != 0 && targetLimit < vettedSigningKeysCount) {
```
Comment (ujenjt) [inline contracts/0.4.24/nos/NodeOperatorsRegistry.sol:1020]: refactor

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol [hunk 30]
```diff
@@ -1047,141 +1085,100 @@ contract NodeOperatorsRegistry is AragonApp, IStakingModule, Versioned {
...
         }
 
-        (address[] memory recipients, uint256[] memory shares) = getRewardsDistribution(sharesToDistribute);
+        (address[] memory recipients, uint256[] memory shares, bool[] memory penalized) =
+            getRewardsDistribution(sharesToDistribute);
 
         distributed = 0;
+
         for (uint256 idx = 0; idx < recipients.length; ++idx) {
             if (shares[idx] == 0) continue;
```
Comment (ujenjt) [inline contracts/0.4.24/nos/NodeOperatorsRegistry.sol:1101]: shares[idx] < 2

Comment (ujenjt) [reply contracts/0.4.24/nos/NodeOperatorsRegistry.sol:1101]: ?

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol (hunk 14 above)
Comment (krogla) [reply contracts/0.4.24/nos/NodeOperatorsRegistry.sol:79]: done

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol (hunk 12 above)
Comment (krogla) [reply contracts/0.4.24/nos/NodeOperatorsRegistry.sol:89]: fixed

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol (hunk 13 above)
Comment (krogla) [reply contracts/0.4.24/nos/NodeOperatorsRegistry.sol:301]: +

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol [hunk 31]
```diff
@@ -309,9 +331,8 @@ contract NodeOperatorsRegistry is AragonApp, IStakingModule, Versioned {
     function setNodeOperatorRewardAddress(uint256 _nodeOperatorId, address _rewardAddress) external {
         _onlyNonZeroAddress(_rewardAddress);
         _onlyExistedNodeOperator(_nodeOperatorId);
-        _authP(SET_NODE_OPERATOR_ADDRESS_ROLE, arr(uint256(_nodeOperatorId), uint256(_rewardAddress)));
+        _authP(ADD_NODE_OPERATOR_ROLE, arr(uint256(_nodeOperatorId), uint256(_rewardAddress)));
 
-        require(_nodeOperators[_nodeOperatorId].rewardAddress != _rewardAddress, "NODE_OPERATOR_ADDRESS_IS_THE_SAME");
```
[the remark this answers is not in the document]
Comment (krogla) [reply contracts/0.4.24/nos/NodeOperatorsRegistry.sol:314]: +

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol (hunk 15 above)
Comment (krogla) [reply contracts/0.4.24/nos/NodeOperatorsRegistry.sol:442]: +
